File: Basic_implementations/nw_linear.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MATCH 1
#define MISMATCH -1
#define GAP -1
#define SEQ_LEN 10000
#define TEST_CASES 25

int max_of_three(int a, int b, int c) {
    if (a >= b && a >= c) return a;
    if (b >= a && b >= c) return b;
    return c;
}

int score(char a, char b) {
    return a == b ? MATCH : MISMATCH;
}

void rev(char *str) {
    int len = strlen(str);
    for (int i = 0; i < len / 2; i++) {
        char tmp = str[i];
        str[i] = str[len - 1 - i];
        str[len - 1 - i] = tmp;
    }
}

char *generate_random_sequence(int len) {
    char *seq = malloc(len + 1);
    char bases[] = {'A', 'C', 'G', 'T'};
    for (int i = 0; i < len; i++) {
        seq[i] = bases[rand() % 4];
    }
    seq[len] = '\0';
    return seq;
}

int validate_alignment(const char *a, const char *b, int expected_score, int *recomputed) {
    int score_ = 0;
    for (int i = 0; a[i] && b[i]; i++) {
        if (a[i] == '_' && b[i] == '_') 
            return 0;
        else if (a[i] == '_' || b[i] == '_') 
            score_ += GAP;
        else 
            score_ += score(a[i], b[i]);
    }
    *recomputed = score_;
    return score_ == expected_score;
}

void needleman_wunsch(char *a, char *b, int test_index) {
    int lenA = strlen(a);
    int lenB = strlen(b);

    int **dp = malloc((lenA + 1) * sizeof(int *));
    char **trace = malloc((lenA + 1) * sizeof(char *));
    for (int i = 0; i <= lenA; i++) {
        dp[i] = malloc((lenB + 1) * sizeof(int));
        trace[i] = malloc((lenB + 1) * sizeof(char));
    }
    /*
    dp[0] → ┌──────────────┐
            │ int, int, ...│ (lenB+1개)
            └──────────────┘
    
    dp[1] → ┌──────────────┐
            │ int, int, ...│
        ...
    */

    dp[0][0] = 0;
    trace[0][0] = 'O';  // origin

    for (int i = 1; i <= lenA; i++) {
        dp[i][0] = i * GAP;
        trace[i][0] = 'U';
    }

    for (int j = 1; j <= lenB; j++) {
        dp[0][j] = j * GAP;
        trace[0][j] = 'L';
    }

    /*
            B  ""   B₁   B₂   B₃   B₄
    A      0   -1   -2   -3   -4
    A₁    -1
    A₂    -2
    A₃    -3


            B  ""   B₁   B₂   B₃   B₄
    A      O    L    L    L    L
    A₁     U
    A₂     U
    A₃     U
    
    */

    for (int i = 1; i <= lenA; i++) {
        for (int j = 1; j <= lenB; j++) {
            int diag = dp[i - 1][j - 1] + score(a[i - 1], b[j - 1]);
            int up = dp[i - 1][j] + GAP;
            int left = dp[i][j - 1] + GAP;

            dp[i][j] = max_of_three(diag, up, left);
            if (dp[i][j] == diag) trace[i][j] = 'D';
            else if (dp[i][j] == up) trace[i][j] = 'U';
            else trace[i][j] = 'L';
        }
    }

    // Traceback
    char *alignedA = malloc(lenA + lenB + 1);
    char *alignedB = malloc(lenA + lenB + 1);
    int ai = 0, bi = 0;
    int i = lenA, j = lenB;

    while (i > 0 || j > 0) {
        if (i > 0 && j > 0 && trace[i][j] == 'D') {
            alignedA[ai++] = a[i - 1];
            alignedB[bi++] = b[j - 1];
            i--; j--;
        } else if (i > 0 && trace[i][j] == 'U') {
            alignedA[ai++] = a[i - 1];
            alignedB[bi++] = '_';
            i--;
        } else if (j > 0 && trace[i][j] == 'L') {
            alignedA[ai++] = '_';
            alignedB[bi++] = b[j - 1];
            j--;
        } else break;
        
    }

    alignedA[ai] = '\0';
    alignedB[bi] = '\0';
    //문자열이므로 끝에 null 삽입
    // 정렬 결과는 뒤집어져 있으니까 다시 rev
    rev(alignedA);
    rev(alignedB);

    // Save result
    char filename[64];
    sprintf(filename, "aligned_result_%d_linear.txt", test_index);
    FILE *fout = fopen(filename, "w");
    fprintf(fout, "[Run %d]\n", test_index);
    fprintf(fout, "Alignment Score: %d\n", dp[lenA][lenB]);
    fprintf(fout, "Aligned A:\n%s\n\n", alignedA);
    fprintf(fout, "Aligned B:\n%s\n", alignedB);
    fclose(fout);
    printf("파일 저장 완료: %s\n", filename);

    for (int i = 0; i <= lenA; i++) {
        free(dp[i]);
        free(trace[i]);
    }
    free(dp); free(trace);
    free(alignedA); free(alignedB);
}

int main() {
    srand(time(NULL));
    for (int t = 1; t <= TEST_CASES; t++) {
        printf("\n==== 테스트 %d ====\n", t);
        char *A = generate_random_sequence(SEQ_LEN);
        char *B = generate_random_sequence(SEQ_LEN);

        clock_t start = clock();
        needleman_wunsch(A, B, t);
        clock_t end = clock();

        double duration = (double)(end - start) / CLOCKS_PER_SEC;
        printf("수행 시간: %.4f초\n", duration);

        free(A); free(B);
    }

    // 검증
    for (int i = 1; i <= TEST_CASES; i++) {
        char filename[64];
        sprintf(filename, "aligned_result_%d_linear.txt", i);
        FILE *f = fopen(filename, "r");
        if (!f) {
            printf("[%d] 파일 없음\n", i);
            continue;
        }

        char line[SEQ_LEN];
        int expected_score = 0;
        char alignedA[SEQ_LEN * 2], alignedB[SEQ_LEN * 2];
        alignedA[0] = alignedB[0] = '\0';

        while (fgets(line, sizeof(line), f)) {
            if (strncmp(line, "Alignment Score:", 16) == 0) {
                sscanf(line + 16, "%d", &expected_score);
            } else if (strncmp(line, "Aligned A:", 10) == 0) {
                fgets(alignedA, sizeof(alignedA), f);
            } else if (strncmp(line, "Aligned B:", 10) == 0) {
                fgets(alignedB, sizeof(alignedB), f);
            }
        }
        fclose(f);
        alignedA[strcspn(alignedA, "\r\n")] = 0;
        alignedB[strcspn(alignedB, "\r\n")] = 0;

        int recomputed = 0;
        int valid = validate_alignment(alignedA, alignedB, expected_score, &recomputed);
        if (valid)
            printf("[%d] 검증 결과: PASS(Recomputed=%d, Expected=%d)\n",i,recomputed,expected_score);
        else
            printf("[%d] 검증 결과: FAIL (Recomputed=%d, Expected=%d)\n", i, recomputed, expected_score);
    }

    return 0;
}

File: Basic_implementations/nw_linear.py
import random
import time
import numpy as np
import os
import multiprocessing as mp

MATCH = 1
MISMATCH = -1
GAP_OPEN = -10
GAP_EXTEND = -1
INF = float('-inf')

def validate_alignment(alignedA, alignedB, expected_score, rel_tol=1e-4):
    score = 0
    gapA = gapB = False
    #gap이 이전에도 있던 건지 새로 열리는건지 확인하는 용도

    #zip으로 두 문자열을 한 글자씩 쌍으로 처리
    for a, b in zip(alignedA, alignedB):
       # 양쪽 다 gap이면 잘못된 정렬
        if a == '_' and b == '_':
            return False, "Invalid: both are gaps"

        elif a == '_':
            if gapA:
                score += GAP_EXTEND
            else:
                score += GAP_OPEN + GAP_EXTEND
                gapA = True
            gapB = False

        elif b == '_':
            if gapB:
                score += GAP_EXTEND
            else:
                score += GAP_OPEN + GAP_EXTEND
                gapB = True
            gapA = False

        else:
            score += MATCH if a == b else MISMATCH
            gapA = gapB = False

    if expected_score != 0:
        relative_error = abs(score - expected_score) / abs(expected_score)
    else:
        relative_error = abs(score - expected_score)  # fallback

    return relative_error <= rel_tol, score

def score(a, b):
    return MATCH if a == b else MISMATCH

def do_traceback(DP, Dx, Dy, trace, traceDx, traceDy, A, B):
    alignedA = []
    alignedB = []
    i, j = len(A), len(B)
    state = 'M'

    while i > 0 or j > 0:
        if state == 'M':
            prev_state = trace[i][j]
            if prev_state == 'M':
                alignedA.append(A[i - 1])
                alignedB.append(B[j - 1])
                i -= 1
                j -= 1
            elif prev_state == 'Dx':
                state = 'Dx'
            elif prev_state == 'Dy':
                state = 'Dy'
        elif state == 'Dx':
            prev_state = traceDx[i][j]
            alignedA.append(A[i - 1])
            alignedB.append('_')
            i -= 1
            state = prev_state
        elif state == 'Dy':
            prev_state = traceDy[i][j]
            alignedA.append('_')
            alignedB.append(B[j - 1])
            j -= 1
            state = prev_state

    return ''.join(reversed(alignedA)), ''.join(reversed(alignedB))





def needleman_cpu(A, B):
    lenA, lenB = len(A), len(B)

    DP = np.full((lenA + 1, lenB + 1), INF)
    Dx = np.full((lenA + 1, lenB + 1), INF)
    Dy = np.full((lenA + 1, lenB + 1), INF)

    trace = [[None] * (lenB + 1) for _ in range(lenA + 1)]
    traceDx = [[None] * (lenB + 1) for _ in range(lenA + 1)]
    traceDy = [[None] * (lenB + 1) for _ in range(lenA + 1)]

    DP[0][0] = 0

    for i in range(1, lenA + 1):
        Dx[i][0] = GAP_OPEN + (i - 1) * GAP_EXTEND
        DP[i][0] = Dx[i][0]
        traceDx[i][0] = 'Dx'
        trace[i][0] = 'Dx'

    for j in range(1, lenB + 1):
        Dy[0][j] = GAP_OPEN + (j - 1) * GAP_EXTEND
        DP[0][j] = Dy[0][j]
        traceDy[0][j] = 'Dy'
        trace[0][j] = 'Dy'

    for i in range(1, lenA + 1):
        for j in range(1, lenB + 1):
            up_ext = Dx[i - 1][j] + GAP_EXTEND
            up_open = DP[i - 1][j] + GAP_OPEN + GAP_EXTEND
            if up_ext >= up_open:
                Dx[i][j] = up_ext
                traceDx[i][j] = 'Dx'
            else:
                Dx[i][j] = up_open
                traceDx[i][j] = 'M'

            left_ext = Dy[i][j - 1] + GAP_EXTEND
            left_open = DP[i][j - 1] + GAP_OPEN + GAP_EXTEND
            if left_ext >= left_open:
                Dy[i][j] = left_ext
                traceDy[i][j] = 'Dy'
            else:
                Dy[i][j] = left_open
                traceDy[i][j] = 'M'

            match = DP[i - 1][j - 1] + score(A[i - 1], B[j - 1])
            if match >= Dx[i][j] and match >= Dy[i][j]:
                DP[i][j] = match
                trace[i][j] = 'M'
            elif Dx[i][j] >= Dy[i][j]:
                DP[i][j] = Dx[i][j]
                trace[i][j] = 'Dx'
            else:
                DP[i][j] = Dy[i][j]
                trace[i][j] = 'Dy'

    return DP, Dx, Dy, trace, traceDx, traceDy

def generate_random_sequence(length):
    return ''.join(random.choices(['A', 'C', 'G', 'T'], k=length))


if __name__=="__main__":
    for run in range(1, 11):
        print(f"\n 테스트 {run} : Needleman-Wunsch 정렬 시작")

        A = generate_random_sequence(10000)
        B = generate_random_sequence(10000)

        start = time.time()
        DP, Dx, Dy, trace, traceDx, traceDy = needleman_cpu(A, B)
        alignedA, alignedB = do_traceback(DP, Dx, Dy, trace, traceDx, traceDy, A, B)
        end = time.time()

        alignment_score = DP[len(A)][len(B)]
        duration = end - start

        print(f"완료 | 점수: {alignment_score} | 시간: {duration:.2f}초")

        # 결과 파일 저장
        filename = f"aligned_result_{run}_twice.txt"
        with open(filename, "w") as f:
            f.write(f"[Run {run}]\n")
            f.write(f"Alignment Score: {alignment_score}\n")
            f.write(f"Execution Time: {duration:.2f} seconds\n\n")
            f.write("Aligned A:\n")
            f.write(alignedA + "\n\n")
            f.write("Aligned B:\n")
            f.write(alignedB + "\n")

        print(f"{filename} 저장 완료")




        for i in range(1, 11):
            filename = f"aligned_result_{i}_twice.txt"
            if not os.path.exists(filename):
                print(f"[{i}]  파일 없음: {filename}")
                continue

            with open(filename, "r") as f:
                lines = f.readlines()

            try:
                expected_score = float(lines[1].strip().split(":")[1])
                a_index = lines.index("Aligned A:\n")
                b_index = lines.index("Aligned B:\n")
            except (ValueError, IndexError):
                print(f"[{i}]  파일 형식 오류")
                continue

            alignedA = lines[a_index + 1].strip()
            alignedB = lines[b_index + 1].strip()

            is_valid, recomputed = validate_alignment(alignedA, alignedB, expected_score, rel_tol=1e-3)
            result = " PASS" if is_valid else f" FAIL (Recalculated={recomputed})"
            print(f"[{i}] 검증 결과: {result}")



